src/client/options.rs

```diff
@@ -185,74 +206,99 @@ impl FromStr for ServerAddress {
 }
 
 impl ServerAddress {
-    /// Parses an address string into a `ServerAddress`.
+    /// Parses an address string into a [`ServerAddress`].
     pub fn parse(address: impl AsRef<str>) -> Result<Self> {
         let address = address.as_ref();
-        // checks if the address is a unix domain socket
+
         #[cfg(unix)]
         {
             if address.ends_with(".sock") {
-                return Ok(ServerAddress::Unix {
-                    path: PathBuf::from(address),
-                });
+                #[cfg(unix)]
+                {
+                    let address =
+                        percent_decode(address, "unix domain sockets must be URL-encoded")?;
+                    return Ok(Self::Unix {
+                        path: PathBuf::from(address),
+                    });
+                }
+                #[cfg(not(unix))]
+                return Err(ErrorKind::InvalidArgument {
+                    message: "unix domain sockets are not supported on this platform",
+                }
+                .into());
             }
         }
-        let mut parts = address.split(':');
-        let hostname = match parts.next() {
-            Some(part) => {
-                if part.is_empty() {
-                    return Err(ErrorKind::InvalidArgument {
-                        message: format!(
-                            "invalid server address: \"{}\"; hostname cannot be empty",
-                            address
-                        ),
-                    }
-                    .into());
+
+        let (hostname, port) = if let Some(ip_literal) = address.strip_prefix("[") {
+            let Some((hostname, port)) = ip_literal.split_once("]") else {
+                return Err(ErrorKind::InvalidArgument {
+                    message: format!(
+                        "invalid server address {}: missing closing ']' in IP literal hostname",
+                        address
+                    ),
                 }
-                part
-            }
-            None => {
+                .into());
+            };
+
+            if let Err(parse_error) = Ipv6Addr::from_str(hostname) {
```

Comment:
was kind of back-and-forth on whether this was valuable, but I think it's worth it because otherwise we may strip the brackets from something that was invalid and attempt to connect to it, e.g. "[localhost]:27017"